### PR TITLE
feat(manager/flux): Update system manifests.

### PR DESCRIPTION
**Pull request by danports**
## Changes

The `flux` manager can now detect and automatically update Flux system manifests.

## Context

Renovating the system manifests allows Flux to update itself automatically once Renovate's changes are merged. Modeled on the Flux [GitHub action](https://github.com/fluxcd/flux2/tree/main/action#automate-flux-updates).

## Documentation

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## Testing

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/manager/flux/index.ts
Comment (viceice) [inline lib/manager/flux/index.ts:23]: We need to extend our buildpack to dynamically install to `sidecar` docker image and to support install mode.

File: lib/manager/flux/index.ts
Comment (danports) [reply lib/manager/flux/index.ts:15]: Alright, I initially started with that approach since I saw it in the `gradle-wrapper` manager, but then I thought maybe `updateDependency` was more relevant since it seemed to be responsible for updating a single file's contents (which is all we need to do for Flux - a single file contains all of the system resources that need to be updated). But if `updateArtifacts` would be better, I'll switch back to that approach.

**Comment by rarkins** [comment]
image

:)

File: lib/manager/flux/index.ts
Comment (danports) [reply lib/manager/flux/index.ts:23]: Yeah, I figured something like that would be necessary - where do I configure that? Maybe it was obvious but I missed it.

**Comment by danports** [comment]
@rarkins I'm too lazy to update dependencies manually - it's easier to open more Renovate PRs... 😂

File: lib/manager/flux/index.ts
Comment (viceice) [reply lib/manager/flux/index.ts:24]: Needs to be done at https://github.com/containerbase/buildpack. You can open a corresponding issue there.

File: lib/manager/flux/index.ts
Comment (viceice) [reply lib/manager/flux/index.ts:16]: Please use `updateArtifacts`

Comment (viceice) [reply lib/manager/flux/index.ts:16]: and add it to `./artifacts.ts` 😉

Comment (danports) [reply lib/manager/flux/index.ts]: Alright, I switched back to the `updateArtifacts` approach - if the general idea is sound now, I'll clean it up, add tests and docs, etc.

Comment (danports) [reply lib/manager/flux/index.ts]: Which approach would you prefer for testing `artifacts.ts`? Should I mock `exec`? Or assume `flux` is installed and run a legit e2e test? I see both approaches in other managers.

File: lib/manager/flux/index.ts
Comment (danports) [reply lib/manager/flux/index.ts]: https://github.com/containerbase/buildpack/pull/250

File: lib/manager/flux/index.ts
Comment (danports) [reply lib/manager/flux/index.ts]: Looks like the CI tests don't run in a container with `flux` installed so I'll go the mock route.

Comment (viceice) [reply lib/manager/flux/index.ts]: Yes, mock exec calls, checkout other managers for samples

**Comment by danports** [comment]
Is there a straightforward way to test `flux` running in a sidecar container before this is merged? I've only tested by manually installing `flux` in my dev container so far.

File: lib/manager/flux/artifacts.ts
Comment (viceice) [inline lib/manager/flux/artifacts.ts:19]:
```suggestion
        image: 'sidecar',
```

File: lib/manager/flux/artifacts.ts
Comment (viceice) [inline lib/manager/flux/artifacts.ts:20]:
```suggestion
```

**Review by viceice — COMMENTED** [review]
Checkout the buildpack.ts file in util/exec for how to install the tool. Then you can run yarn start with binarySource=docker

**Comment by danports** [comment]
Verified that everything works when `binarySource` is `docker`. 🥳 I think we are ready for 🚀 here unless there are some more nits to address.

File: lib/manager/flux/artifacts.ts
Comment (rarkins) [inline lib/manager/flux/artifacts.ts:11]: Can you explain this requirement/limitation that "first updated dep must have a new version"? E.g. is it your intent that the system manifest is upgraded on its own and not mixed with other upgrades of dependencies?

File: lib/manager/flux/artifacts.ts
Comment (rarkins) [inline lib/manager/flux/artifacts.ts:34]: The only file which changes the manifest file itself?

File: lib/manager/flux/extract.ts
Comment (rarkins) [inline lib/manager/flux/extract.ts:15]: You test whether it's a system manifest based on file name alone, so this could potentially not match if the file contents are not as expected?

File: lib/manager/flux/index.ts
Comment (viceice) [inline lib/manager/flux/index.ts:9]: Move to `./common.ts`, it's private API

File: lib/manager/flux/index.ts
Comment (viceice) [inline lib/manager/flux/index.ts:8]: Use `regex` function from `utils/regex`.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/manager/flux/artifacts.ts
Comment (danports) [reply lib/manager/flux/artifacts.ts:11]: This logic is based on how system manifests are treated in `extract.ts`. Essentially, we treat a system manifest file as a single unit that has only one `github-releases` dependency on Flux itself. In other words, if `isSystemManifest(packageFileName)` returns `true` here, then `updatedDeps.length` will be `1`. This `updatedDeps[0]?.newVersion` check is only here as a defensive measure and should never actually fail.

File: lib/manager/flux/artifacts.ts
Comment (danports) [reply lib/manager/flux/artifacts.ts:34]: That's correct.

File: lib/manager/flux/extract.ts
Comment (danports) [reply lib/manager/flux/extract.ts:15]: True, I'll add a test for this.

**Review by rarkins — DISMISSED** [review]

**Comment by danports** [comment]
@viceice Any additional feedback for me to address?

**Review by viceice — APPROVED** [review]
LGTM, waiting for docs review 🙃

File: lib/manager/flux/readme.md
Comment (viceice) [inline lib/manager/flux/readme.md:1]: We should note, that this currently isn't supported on renovate full docker image

 - renovatebot/docker-renovate-full#406

**Review by viceice — DISMISSED** [review]

File: lib/manager/flux/readme.md
Comment (rarkins) [inline lib/manager/flux/readme.md:5]:
```suggestion

Updating system manifests requires that the `flux` tool be pre-installed or you are running a Docker image based off [containerbase/buildpack](https://github.com/containerbase/buildpack), such as the official Renovate images, and have `binarySource=install` configured.

```

Comment (danports) [reply lib/manager/flux/readme.md:5]: Added this a bit lower in the docs (to match the order of the features described at the top of the file) and separated the two options into bullet points to avoid any doubt about whether `binarySource=install` is required for both options or just the second option.

File: lib/manager/flux/readme.md
Comment (danports) [reply lib/manager/flux/readme.md:1]: 👍 I think the note I just added to the docs should cover this.

**Review by rarkins — APPROVED** [review]

File: lib/manager/flux/artifacts.ts
Comment (rarkins) [inline lib/manager/flux/artifacts.ts:35]: This unfortunately also captures the output from `install-tool`

**Pull request merged**
